### PR TITLE
Code Health - Add NPN API documentation

### PR DESCRIPTION
**Pull request by jon-oracle**
##### Checklist
- [x] documentation is added or updated

##### Description of change
Added documentation for Next Protocol Negotiation APIs.
[skip ci]

### CODE REVIEW

File: doc/man3/SSL_CTX_set_alpn_select_cb.pod [hunk 1]
```diff
@@ -21,13 +22,21 @@ protocol negotiation (ALPN)
                                             const unsigned char *in,
                                             unsigned int inlen,
                                             void *arg), void *arg);
+ void SSL_CTX_set_next_protos_advertised_cb(SSL_CTX *ctx,
+                                   SSL_CTX_npn_advertised_cb_func cb,
+                                   void *arg);
+ void SSL_CTX_set_next_proto_select_cb(SSL_CTX *ctx,
+                               SSL_CTX_npn_select_cb_func cb,
```
Comment (kroeckx) [inline doc/man3/SSL_CTX_set_alpn_select_cb.pod]: I don't see the SSL_CTX_npn_select_cb_func anywhere, and so talking about parameters you give to that is hard to follow.

File: doc/man3/SSL_CTX_set_alpn_select_cb.pod [hunk 2]
```diff
@@ -119,14 +154,21 @@ ALPN protocol not selected.
 
 =back
 
+The SSL_CTX_set_next_proto_select_cb() callback should return
+B<SSL_TLSEXT_ERR_OK> if successful. Any other value is fatal to the connection.
```
Comment (kroeckx) [inline doc/man3/SSL_CTX_set_alpn_select_cb.pod]: Do you mean the function set with set with SSL_CTX_set_next_proto_select_cb(), so `cb`? Since this is not the return value of SSL_CTX_set_next_proto_select_cb() itself, I think it should get documented where you describe the callback.

**Comment by jon-oracle** [comment]
Tidied up the description a bit. Agree with your point re ALPN. I've done some work in this area so just thought I'd add what was missing.

**Comment by richsalz** [comment]
also, i assume find-doc-nits comes up clean here :)

**Comment by jon-oracle** [comment]
Yeah find-doc-nits runs ok ;)

**Comment by richsalz** [comment]
@kroeckx are your concerns addressed?

File: doc/man3/SSL_CTX_set_alpn_select_cb.pod [hunk 3]
```diff
@@ -59,10 +77,36 @@ B<client>, so it should be copied immediately. If no match is found, the first
...
 function can also be used in the NPN callback.
 
+SSL_CTX_set_next_proto_select_cb() sets a callback B<cb> that is called when a
+client needs to select a protocol from the server's provided list. B<out>
+must be set to point to the selected protocol (which may be within B<in>).
+The length of the protocol name must be written into B<outlen>. The
+server's advertised protocols are provided in B<in> and B<inlen>. The
+callback can assume that B<in> is syntactically valid. The client must
+select a protocol. It is fatal to the connection if this callback returns
+a value other than B<SSL_TLSEXT_ERR_OK>.
```
Comment (kroeckx) [inline doc/man3/SSL_CTX_set_alpn_select_cb.pod]: It does not mention anything about arg.

File: doc/man3/SSL_CTX_set_alpn_select_cb.pod [hunk 4]
```diff
@@ -59,10 +77,36 @@ B<client>, so it should be copied immediately. If no match is found, the first
...
 SSL_get0_alpn_selected() returns a pointer to the selected protocol in B<data>
 with length B<len>. It is not NUL-terminated. B<data> is set to NULL and B<len>
 is set to 0 if no protocol has been selected. B<data> must not be freed.
 
+SSL_get0_next_proto_negotiated() sets B<data> and B<len> to point to the
+client's requested protocol for this connection and returns 0. If the
+client didn't request any protocol, then B<data> is set to NULL. Note that
+the client can request any protocol it chooses. The value returned from
+this function need not be a member of the list of supported protocols
+provided by the callback.
```
Comment (kroeckx) [inline doc/man3/SSL_CTX_set_alpn_select_cb.pod]: What happens in case NPN wasn't enabled? Can it ever return something other than 0?

File: doc/man3/SSL_CTX_set_alpn_select_cb.pod [hunk 5]
```diff
@@ -119,14 +163,21 @@ ALPN protocol not selected.
 
 =back
 
+The callback set using SSL_CTX_set_next_proto_select_cb() should return
+B<SSL_TLSEXT_ERR_OK> if successful. Any other value is fatal to the connection.
```
Comment (kroeckx) [inline doc/man3/SSL_CTX_set_alpn_select_cb.pod]: Can it return something other to not sent an error, like I think it does in case no cb is present? Also see #2570.

**Comment by richsalz** [comment]
87b8149 in master and 48e017b in 1.1.0  Thanks!
